### PR TITLE
Consistent builder API

### PR DESCRIPTION
**Pull request by ltratt**
This PR homogenises lrlex and lrpar to use a consistent, and "traditional", builder API for compile-time building. It might be easiest to start with the summary document changes in https://github.com/softdevteam/grmtools/commit/73597c1e7c7033541ce64c92dfbe4f5f1e9e9256, which shows that the changes aren't very scary, before working through the individual commits.

https://github.com/softdevteam/grmtools/commit/6f53c8c7d43a96194c43ceb0538d8f0c200b231e moves lrlex from a non-consuming `process` function to a consuming `build` function that returns a `CTParser` struct. https://github.com/softdevteam/grmtools/commit/43c6d46318f415649dad6ea87c08be95965f0ea7 moves lrlex from an already-consuming `process` to a renamed `build` function. While doing this, another long-running inconsistency got my goat: lrlpar has `CTParserBuilder` (for "compile-time") while lrlex used `LexerBuilder`. https://github.com/softdevteam/grmtools/commit/18f852806bd4d2b2a405ce8e416cdc8d65af9586 renames lrlex's struct appropriately. https://github.com/softdevteam/grmtools/commit/3dda7c266fcbae4aa7b5f1bd7940ae4f32d3c7f0 then adjusts lrlex to return a `CTLexer`, mirroring lrpar, and giving us more flexibility for the future.

A fun thing about these changes is that through careful use of `#deprecated` and function and type aliases, pretty much all old code continues to compile and run correctly with these changes (only code that used a function marked as "unstable" fails to compile), albeit producing warnings that the user should update their code. Still, those warnings will be a bit annoying, so I think this PR strongly suggests the next grmtools release should be 0.11.0.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -1,6 +1,41 @@
 # grmtools 0.11.0 (XXXX-YY-ZZ)
 
-## APIs deprecated
+## API deprecation summary
+
+* Several of the functions / structs surrounding the compile-time construction
+  of grammars have changed: more details are given below, but in most cases
+  a `build.rs` that looks as follows:
```
Comment (ptersilie) [inline CHANGES.md:7]: ~~that~~

File: CHANGES.md [hunk 2]
```diff
@@ -1,6 +1,41 @@
...
+    }
+  ```
+  can be changed to:
+  ```rust
+    use lrlex::CTLexerBuilder;
+    use lrpar::CTParserBuilder;
+
+    fn main() -> Result<(), Box<dyn std::error::Error>> {
+        let cp = CTParserBuilder::<u8>::new_with_storaget()
+            .grammar_in_src_dir("config.y")?
```
Comment (ptersilie) [inline CHANGES.md:28]: do we not need a `process()` after this?

Comment (ptersilie) [reply CHANGES.md:28]: Oh I see `process` has been replaced in favour of `build`.

Comment (ltratt) [reply CHANGES.md:28]: Yep, to make clear that `build` consumes the builder.

**Comment by ptersilie** [comment]
Some typos in commit messages:
43c6d46: Since ~~build~~ `process` already consumes the builder, this is a fairly easy
change.
18f8528: It ~~is~~ has been confusing me...

**Comment by ptersilie** [comment]
Few commits. Rest looks good.

**Comment by ltratt** [comment]
Shall I force push an update?

**Comment by ptersilie** [comment]
Once you've addressed: https://github.com/softdevteam/grmtools/pull/249#discussion_r709123534

File: CHANGES.md (hunk 1 above)
Comment (ltratt) [reply CHANGES.md:7]: I think "that" is correct here, because we're making clear that we're only talking about `build.rs` files which follow this pattern.

**Comment by ltratt** [comment]
> Once you've addressed: #249 (comment)

Oops!

File: CHANGES.md (hunk 1 above)
Comment (ptersilie) [reply CHANGES.md:7]: The sentence still sounds weird: more details are given below, but in most cases a `build.rs` that looks as follows.

What's the "that" for?

Comment (ltratt) [reply CHANGES.md:7]: There's a follow-on sentence after the code block so it's effectively "in most cases
  a `build.rs` that looks as follows <...> can be changed to <...>".

Comment (ptersilie) [reply CHANGES.md:7]: Ooooohhhhh! Ignore me. :-)

**Comment by ptersilie** [comment]
Please squash.

**Comment by ltratt** [comment]
Force pushed updated commit messages.

**Pull request merged**